### PR TITLE
IGNITE-13310 Adds tracing of SQL queries.

### PR DESCRIPTION
**Pull request by ololo3000**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/tracing/TraceableIterator.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+     * @param iter Iterator to which all calls will be delegated.
+     */
+    public TraceableIterator(Iterator<T> iter) {
+        this.iter = iter;
+        this.span = MTC.span();
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean hasNext() {
+        MTC.supportInitial(span);
```
Comment (sanpwc) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/tracing/TraceableIterator.java:44]: This seems a little bit confusing. Abstracting from the current format of use, on every iteration we reset MTC.span with what we get from MTC.span() within constructor. Why we need this?

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/dml/DmlUtils.java [hunk 2]
```diff
@@ -194,10 +197,16 @@ private static long dmlDoInsert(UpdatePlan plan, Iterable<List<?>> cursor, int p
...
             IgniteBiTuple t = plan.processRow(cursor.iterator().next());
 
-            if (cctx.cache().putIfAbsent(t.getKey(), t.getValue()))
-                return 1;
-            else
-                throw new TransactionDuplicateKeyException("Duplicate key during INSERT [key=" + t.getKey() + ']');
+            try (
+                MTC.TraceSurroundings ignored = MTC.support(cctx.kernalContext().tracing()
+                    .create(SQL_CACHE_UPDATE, MTC.span())
+                    .addTag(SQL_CACHE_UPDATES, () -> Integer.toString(1)))
```
Comment (sanpwc) [inline modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/dml/DmlUtils.java:203]: Constant?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/tracing/TraceableIterator.java (hunk 1 above)
Comment (ololo3000) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/tracing/TraceableIterator.java:44]: The main reason to use `MTC.supportInitial(span);` is that methods `hasNext` and `next` can be called from user threads for which MTC.span is not set. During this methods execution important spans produce and they should be part of span tree that represents SQL query to which current iterator belongs to. So if we don't use `MTC.supportInitial(span);` this spans will not be inherited from the query root span.

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/dml/DmlUtils.java (hunk 2 above)
Comment (ololo3000) [reply modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/dml/DmlUtils.java]: Thanks. Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/tracing/TraceableIterator.java (hunk 1 above)
Comment (sanpwc) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/tracing/TraceableIterator.java:44]: Well, TraceableIterator still looks confusing to me, however, cause it's further use is unclear I'm ok with it.

**Pull request merged**